### PR TITLE
numpy 2.0: fix slicing by uint64 array

### PR DESCRIPTION
**Pull request by crusaderky**
- Part of #10593

### CODE REVIEW

File: dask/array/slicing.py [hunk 1]
```diff
@@ -561,7 +561,7 @@ def slicing_plan(chunks, index):
         index = np.asanyarray(index)
     cum_chunks = cached_cumsum(chunks)
 
-    cum_chunks = asarray_safe(cum_chunks, like=index)
+    cum_chunks = asarray_safe(cum_chunks, like=index, dtype=index.dtype)
```
Comment (crusaderky) [inline dask/array/slicing.py:564]: cum_chunks was int64 before (in numpy 1 too); it now changes to uint64.
This fixes the dtype arithmetics on line 581 below, which used to subtract uint64 - int64, which in numpy 1 returns uint64 but in numpy 2 returns float.

File: dask/array/slicing.py [hunk 2]
```diff
@@ -634,8 +634,6 @@ def take(outname, inname, chunks, index, itemsize, axis=0):
             PerformanceWarning,
             stacklevel=6,
         )
-    if not is_arraylike(index):
-        index = np.asarray(index)
```
Comment (crusaderky) [inline dask/array/slicing.py:638]: dead code

File: dask/array/slicing.py [hunk 3]
```diff
@@ -559,9 +559,14 @@ def slicing_plan(chunks, index):
...
         index = np.asanyarray(index)
-    cum_chunks = cached_cumsum(chunks)
 
-    cum_chunks = asarray_safe(cum_chunks, like=index)
+    cum_chunks_tup = cached_cumsum(chunks)
+    cum_chunks = asarray_safe(cum_chunks_tup, like=index)
+    if cum_chunks.dtype.kind != "f":  # Don't cast NaN chunks to int
+        # This is important when index.dtype=uint64 (or uint32 on 32-bit hosts) to
+        # prevent accidental automatic casting during `index - cum_chunks` below
+        cum_chunks = cum_chunks.astype(index.dtype)
```
Comment (crusaderky) [inline dask/array/slicing.py:568]: cum_chunks was int64 before (in numpy 1 too); this PR changes it to uint64 (or whatever the end user passed as a slicer).
This fixes the dtype arithmetics on line 586 below, which used to subtract uint64 - int64, which in numpy 1 returns uint64 but in numpy 2 returns float.

**Review by hendrikmakait — APPROVED** [review]
Thanks, @crusaderky!

**Pull request merged**